File: VMusicPlayerServer/src/main/kotlin/com/github/rain1208/vmusicplayerserver/domain/models/Song.kt
package com.github.rain1208.vmusicplayerserver.domain.models

import com.github.rain1208.vmusicplayerserver.repository.dto.MusicSourceDto
import com.github.rain1208.vmusicplayerserver.repository.dto.SingerDto
import com.github.rain1208.vmusicplayerserver.repository.dto.SongDto
import java.time.Duration

data class Song(
    val songId: String,
    val title: String,
    val artist: String?,
    val startAt: Duration,
    val endAt: Duration,
    val source: MusicSourceDto,
    val singers: List<SingerDto>,
) {
    fun toSongDto(): SongDto {
        return SongDto(
            songId = songId,
            title = title,
            artist = artist,
            startAt = startAt,
            endAt = endAt,
            sourceId = source.sourceId,
            singerIds = singers.map { it.singerId }
        )
    }
}

File: VMusicPlayerApp/types/song.ts
export interface Singer {
  singerId: string
  singerName: string
  youtubeURL?: string
  twitterURL?: string
}

export interface MusicSource {
  sourceId: string
  title: string
  url: string
  uploadDate: string
  thumbnailUrl: string
  sourceType: string
}

export interface Song {
  songId: string
  title: string
  artist?: string
  startAt: string
  endAt: string
  source: MusicSource
  singers: Singer[]
}

